File: Graphs/BFSearch/MangoSellerProblem/MangoSellerSearch.py
from collections import deque

# Define o grafo como um dicionário de listas
graph = {
    "you": ["alice", "bob", "claire"],
    "bob": ["anuj", "peggy"],
    "alice": ["peggy"],
    "claire": ["thom", "jonny"],
    "anuj": [],
    "peggy": [],
    "thom": [],
    "jonny": []
}

def search(name):
    queue = deque(graph[name])
    checked = []

    while queue:
        person = queue.popleft()

        if person not in checked:
            if is_mango_seller(person):
                print(person + " is a mango seller!")
                return True
            else:
                queue.extend(graph[person])
                checked.append(person)

    return False

def is_mango_seller(person):
    return person.endswith("m")

if __name__ == "__main__":
    search("you")

File: Graphs/BFSearch/MangoSellerProblem/MangoSellerSearch.go
package main

import (
	"fmt"
)

var graph = map[string][]string{
	"you":    {"alice", "bob", "claire"},
	"bob":    {"anuj", "peggy"},
	"alice":  {"peggy"},
	"claire": {"thom", "jonny"},
	"anuj":   {},
	"peggy":  {},
	"thom":   {},
	"jonny":  {},
}

func search(name string) bool {
	queue := append([]string{}, graph[name]...)
	checked := make(map[string]bool)

	for len(queue) > 0 {
		person := queue[0]
		queue = queue[1:]

		if !checked[person] {
			if isMangoSeller(person) {
				fmt.Println(person, "is a mango seller!")
				return true
			} else {
				queue = append(queue, graph[person]...)
				checked[person] = true
			}
		}
	}
	return false
}

func isMangoSeller(person string) bool {
	return len(person) > 0 && person[len(person)-1] == 'm'
}

func main() {
	search("you")
}
